- Expand SamLocusAndReferenceIterator tests to cover BAMs and CRAMs.

yfarjoun opened the pull request:
(Sorry about the premature push into master earlier)

There seems to be a lack of testing of SamLocusAndReferenceIterator on cram and bam files, which lead one of our contributors to submit #1483 which seems to be solving a problem that doesn't exist. I hope that this PR will provide an example of how to properly create a SamLocusAndReferenceIterator in a way that is agnostic of the type of file (sam/bam/cram) at hand.

Commit:
- Expand SamLocusAndReferenceIterator tests to cover BAMs and CRAMs.

lbergelson commented on src/main/java/htsjdk/variant/vcf/SamFileUtils.java:
```diff
@@ -0,0 +1,63 @@
+ * THE SOFTWARE.
+ */
+
+package htsjdk.variant.vcf;
```
Looks like this is the wrong package for something called "SamFileUtils"

lbergelson replied:
It would probably be better to put this method in htsjdk.samtools.util.TestUtil instead of a new class.

lbergelson commented on src/main/java/htsjdk/variant/vcf/SamFileUtils.java:
```diff
@@ -0,0 +1,63 @@
+        final File out = File.createTempFile(prefix, suffix);
+        out.deleteOnExit();
+        String indexFileExtension = null;
+        if (suffix.endsWith("vcf.gz")) {
```
these should probably reference the constants in FileExtensions

lbergelson replied:
This would probably be better placed in something like htsjdk.samtools.util.TestUtil instead of a new class

lbergelson commented on src/test/java/htsjdk/samtools/reference/SamLocusAndReferenceIteratorTest.java:
```diff
@@ -1,26 +1,87 @@
+        };
+    }
+
+    @BeforeSuite
```
Could we change this from `@BeforeSuite` to `@BeforeClass` ?  That will limit the chance of confusing things happening when this setup method is unexpectedly run before unrelated tests. (https://testng.org/doc/documentation-main.html#annotations)

lbergelson reviewed (changes requested):
@yfarjoun Thanks for doing this.  I have a few comments then you're good to merge.

Commit:
- respond to review comments
- In FileExtension the Bam index is called BAI_INDEX which is different from all the others (CRAM_INDEX, VCF_INDEX, etc) so I added BAM_INDEX, and deprecated BAI_INDEX

lbergelson commented on src/main/java/htsjdk/samtools/util/FileExtensions.java:
```diff
@@ -53,7 +53,12 @@
     public static final String SAM = ".sam";
     public static final String BAM = ".bam";
-    public static final String BAI_INDEX = ".bai";
+    public static final String BAM_INDEX = ".bai";
```
I think this was deliberately called the BAI index in able to differentiate it from BAM CSI index.

yfarjoun replied:
hmmm. good point. I guess the name should be `BAM_BAI_INDEX` and `BAM_CSI_INDEX`

lbergelson commented on src/main/java/htsjdk/samtools/util/TestUtil.java:
```diff
@@ -89,6 +90,35 @@ public static File getTempDirecory(final String prefix, final String suffix) {
+        }
+
+        if (indexFileExtension != null) {
+            final File indexOut = new File(out.getAbsolutePath() + indexFileExtension);
```
Oh, I didn't think about this, but don't we need to be careful about extension vs replacement here?  Don't some of these follow replacement rules instead of additive rules.  I.e. bam.bai is ok, but it's just .crai instead of .cram.crai?

yfarjoun replied:
I guess we need objects that generate index_names from file names...

lbergelson replied:
We have them, they're just not centralized.

yfarjoun replied:
where?

yfarjoun replied:
they are impossible to find!

Commit:
- more response and refactor

Commit:
change to reflect reality + comment

yfarjoun commented:
back to you @lbergelson
